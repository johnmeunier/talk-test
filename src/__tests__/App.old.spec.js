import renderer from "react-test-renderer";
import { shallow, configure } from "enzyme";
import Adapter from "enzyme-adapter-react-16";

import Container, {
  View,
  toggleStatus,
  handleAddTask,
  filter,
  handleFilterStatus,
} from "../App.old";

describe("App old way", () => {
  beforeAll(() => {
    configure({ adapter: new Adapter() });
  });
  describe("testing pure functions", () => {
    test("toggleStatus", () => {
      const mockSetTask = jest.fn();

      toggleStatus(mockSetTask)(2);
      const callback = mockSetTask.mock.calls[0][0];

      expect(
        callback([
          { completed: false },
          { completed: false },
          { completed: false },
          { completed: false },
        ])
      ).toStrictEqual([
        { completed: false },
        { completed: false },
        { completed: true },
        { completed: false },
      ]);
    });

    test("handleAddTask should add task to the list", () => {
      const mockSetTasks = jest.fn();

      handleAddTask(mockSetTasks, jest.fn())("new item");
      const callback = mockSetTasks.mock.calls[0][0];

      expect(callback([{ value: 1 }, { foo: 2 }])).toStrictEqual([
        { value: 1 },
        { foo: 2 },
        { id: 3, label: "new item", completed: false },
      ]);
    });

    test("handleAddTask should reset newTodo to empty", () => {
      const mockSetNewTodo = jest.fn();

      handleAddTask(jest.fn(), mockSetNewTodo)("new item");
      const callback = mockSetNewTodo.mock.calls[0][0];

      expect(callback()).toEqual("");
    });

    const mockData = [
      {
        id: 0,
        label: "active 1",
        completed: false,
      },
      {
        id: 1,
        label: "completed 1",
        completed: true,
      },
      {
        id: 2,
        label: "active 2",
        completed: false,
      },
      {
        id: 3,
        label: "completed 2",
        completed: true,
      },
    ];
    test.each([
      ["all", mockData],
      [
        "completed",
        [
          {
            id: 1,
            label: "completed 1",
            completed: true,
          },
          {
            id: 3,
            label: "completed 2",
            completed: true,
          },
        ],
      ],
      [
        "active",
        [
          {
            id: 0,
            label: "active 1",
            completed: false,
          },
          {
            id: 2,
            label: "active 2",
            completed: false,
          },
        ],
      ],
      ["unknown", undefined],
    ])(
      "filter should keep items with matching status (%p)",
      (filterStatus, expectedTasks) => {
        expect(filter({ tasks: mockData, filterStatus })).toStrictEqual(
          expectedTasks
        );
      }
    );

    test("handleFilterStatus", () => {
      const mockSetFilterStatus = jest.fn();
      handleFilterStatus(mockSetFilterStatus)({ target: { value: 55 } });
      expect(mockSetFilterStatus).toHaveBeenCalledWith(55);
    });
  });

  describe("container", () => {
    test("it should render without crashing", () => {
      const tree = renderer.create(<Container />).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });

  describe("view component", () => {
    const defaultProps = {
      filteredTasks: [],
      filterStatus: "all",
      newTodo: "",
      toggleStatus: jest.fn(),
      handleFilterStatus: jest.fn(),
      setNewTodo: jest.fn(),
      handleAddTask: jest.fn(),
    };

    test("initial state", () => {
      const tree = renderer.create(<View {...defaultProps} />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    test("with tasks", () => {
      const props = {
        ...defaultProps,
        filteredTasks: [
          { id: 1, label: "first task", completed: false },
          { id: 2, label: "first task", completed: true },
        ],
      };
      const tree = renderer.create(<View {...props} />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    test("with new todo", () => {
      const props = {
        ...defaultProps,
        newTodo: "new task",
      };
      const tree = renderer.create(<View {...props} />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    test("filter click should call handleFilterStatus", () => {
      const wrapper = shallow(<View {...defaultProps} />);
      const input = wrapper.find("input[type='radio']").at(0);

      input.simulate("change", { checked: true });
      expect(defaultProps.handleFilterStatus).toHaveBeenCalledWith({
        checked: true,
      });
    });

    test("toggle status of item should call toggleStatus", () => {
      const props = {
        ...defaultProps,
        filteredTasks: [
          { id: 1, label: "first task", completed: false },
          { id: 2, label: "first task", completed: true },
        ],
      };

      const wrapper = shallow(<View {...props} />);
      const input = wrapper.find("li.task").at(1);

      input.simulate("click");
      expect(defaultProps.toggleStatus).toHaveBeenCalledWith(1);
    });

    test("typing in new todo input should call setNewTodo", () => {
      const wrapper = shallow(<View {...defaultProps} />);
      const input = wrapper.find("input[type='text']");

      input.simulate("change", { target: { value: "newValue" } });
      expect(defaultProps.setNewTodo).toHaveBeenCalledWith("newValue");
    });

    test("click the add button should call handleAddTask", () => {
      const props = { ...defaultProps, newTodo: "new value to set" };
      const wrapper = shallow(<View {...props} />);
      const input = wrapper.find("button");

      input.simulate("click", { target: { value: "newValue" } });
      expect(props.handleAddTask).toHaveBeenCalledWith(props.newTodo);
    });
  });
});
