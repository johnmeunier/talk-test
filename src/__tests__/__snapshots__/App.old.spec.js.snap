// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App old way container it should render without crashing 1`] = `
<div
  className="App"
>
  <div
    className="filter__container"
  >
    <div
      className="filter"
    >
      <label>
        completed
        <input
          checked={false}
          name="status"
          onChange={[Function]}
          type="radio"
          value="completed"
        />
      </label>
      <label>
        active
        <input
          checked={false}
          name="status"
          onChange={[Function]}
          type="radio"
          value="active"
        />
      </label>
      <label>
        all
        <input
          checked={true}
          name="status"
          onChange={[Function]}
          type="radio"
          value="all"
        />
      </label>
    </div>
  </div>
  <div
    className="task__container"
  >
    <ul />
  </div>
  <label>
    Add
    <input
      onChange={[Function]}
      type="text"
      value=""
    />
  </label>
  <button
    onClick={[Function]}
    type="button"
  >
    Add
  </button>
</div>
`;

exports[`App old way view component initial state 1`] = `
<div
  className="App"
>
  <div
    className="filter__container"
  >
    <div
      className="filter"
    >
      <label>
        completed
        <input
          checked={false}
          name="status"
          onChange={[MockFunction]}
          type="radio"
          value="completed"
        />
      </label>
      <label>
        active
        <input
          checked={false}
          name="status"
          onChange={[MockFunction]}
          type="radio"
          value="active"
        />
      </label>
      <label>
        all
        <input
          checked={true}
          name="status"
          onChange={[MockFunction]}
          type="radio"
          value="all"
        />
      </label>
    </div>
  </div>
  <div
    className="task__container"
  >
    <ul />
  </div>
  <label>
    Add
    <input
      onChange={[Function]}
      type="text"
      value=""
    />
  </label>
  <button
    onClick={[Function]}
    type="button"
  >
    Add
  </button>
</div>
`;

exports[`App old way view component with new todo 1`] = `
<div
  className="App"
>
  <div
    className="filter__container"
  >
    <div
      className="filter"
    >
      <label>
        completed
        <input
          checked={false}
          name="status"
          onChange={[MockFunction]}
          type="radio"
          value="completed"
        />
      </label>
      <label>
        active
        <input
          checked={false}
          name="status"
          onChange={[MockFunction]}
          type="radio"
          value="active"
        />
      </label>
      <label>
        all
        <input
          checked={true}
          name="status"
          onChange={[MockFunction]}
          type="radio"
          value="all"
        />
      </label>
    </div>
  </div>
  <div
    className="task__container"
  >
    <ul />
  </div>
  <label>
    Add
    <input
      onChange={[Function]}
      type="text"
      value="new task"
    />
  </label>
  <button
    onClick={[Function]}
    type="button"
  >
    Add
  </button>
</div>
`;

exports[`App old way view component with tasks 1`] = `
<div
  className="App"
>
  <div
    className="filter__container"
  >
    <div
      className="filter"
    >
      <label>
        completed
        <input
          checked={false}
          name="status"
          onChange={[MockFunction]}
          type="radio"
          value="completed"
        />
      </label>
      <label>
        active
        <input
          checked={false}
          name="status"
          onChange={[MockFunction]}
          type="radio"
          value="active"
        />
      </label>
      <label>
        all
        <input
          checked={true}
          name="status"
          onChange={[MockFunction]}
          type="radio"
          value="all"
        />
      </label>
    </div>
  </div>
  <div
    className="task__container"
  >
    <ul>
      <li
        className="task"
        onClick={[Function]}
      >
        first task
         
        <span
          className="status-icon "
        >
          <svg
            role="img"
            viewBox="0 0 96 96"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title>
              Active
            </title>
            <path
              d="M1 1v94h94V1H1zM91.1 91.1H4.9V4.9h86.1V91.1z"
            />
            <path
              className="completed-check"
              d="m 22.939024,46.886041 18.480655,18.0492 31.424305,-33.078215 v 0"
              fill="none"
            />
          </svg>
        </span>
      </li>
      <li
        className="task"
        onClick={[Function]}
      >
        first task
         
        <span
          className="status-icon status-icon--completed"
        >
          <svg
            role="img"
            viewBox="0 0 96 96"
            xmlns="http://www.w3.org/2000/svg"
          >
            <title>
              Completed
            </title>
            <path
              d="M1 1v94h94V1H1zM91.1 91.1H4.9V4.9h86.1V91.1z"
            />
            <path
              className="completed-check"
              d="m 22.939024,46.886041 18.480655,18.0492 31.424305,-33.078215 v 0"
              fill="none"
            />
          </svg>
        </span>
      </li>
    </ul>
  </div>
  <label>
    Add
    <input
      onChange={[Function]}
      type="text"
      value=""
    />
  </label>
  <button
    onClick={[Function]}
    type="button"
  >
    Add
  </button>
</div>
`;
